### ISSUE TITLE
Exit code 137 when running tennis_question_typology.py

### ISSUE THREAD

**Issue by lindieee** [CornellNLP/Cornell-Conversational-Analysis-Toolkit#19]
Hi,
when running tennis_question_typology.py in PyCharm I always get the error message "Process finished with exit code 137. Interrupted by signal 9: SIGKILL." The error occurs when the list "spacy_objs" is dumped with pickle and written to spacy.pk: 

questionTypology.py - class MotifsExtractor - def spacify - 
with open(outfile_name + '.pk', 'wb') as f:
    pickle.dump(spacy_objs, f)  

Exit code 137 seems to be an out of memory problem. But I have 16 GB RAM and I already increased the max memory size that PyCharm is allowed to use in the PyCharm settings. So I do not see how memory can be a problem here. 

Is it still possible that the corpus file is just too big?

**Comment by jpwchang** [comment]
We've been looking into this problem, and there are two possible reasons for what you're experiencing, both of which stem from the same root cause. The first possibility is that spacy_objs is too large for pickle to handle, causing in to crash (as reported [here](https://stackoverflow.com/questions/29704139/pickle-in-python3-doesnt-work-for-large-data-saving/29704623#29704623)). The second possibility is that you actually are just hitting the memory limit on your machine - but don't worry, the corpus file isn't actually 16 GB in size!

Regardless of which issue is leading to the error you're seeing, the root cause is in the SpaCy dependency. SpaCy 2.0, which we upgraded to last year, contains a lot of really useful changes but also introduces an issue with memory usage on large datasets (see [this github issue](https://github.com/explosion/spaCy/issues/1600)). This causes the spacy_objs object to eat much more memory than it really needs to, and by extension causes the pickled object to be much larger than needed. Thankfully, there is a workaround for this, which we have implemented and will be rolling out in the 2.0 release.

**Comment by lindieee** [comment]
Thank you very much for your response! I followed your links and tried the ideas given there as far as possible but none of it did work for me.

 

Is it perhaps possible to get some kind of pre-version of your 2.0 release? The reason I am asking is that I really would like to use the convokit question typology part in a student project I am doing right now and at the moment I do not know how to fix this error.

 

 

Gesendet: Donnerstag, 24. Januar 2019 um 16:17 Uhr
Von: "Jonathan Chang" <notifications@github.com>
An: CornellNLP/Cornell-Conversational-Analysis-Toolkit <Cornell-Conversational-Analysis-Toolkit@noreply.github.com>
Cc: lindieee <arettig@gmx.de>, Author <author@noreply.github.com>
Betreff: Re: [CornellNLP/Cornell-Conversational-Analysis-Toolkit] Exit code 137 when running tennis_question_typology.py (#19)


We've been looking into this problem, and there are two possible reasons for what you're experiencing, both of which stem from the same root cause. The first possibility is that spacy_objs is too large for pickle to handle, causing in to crash (as reported here). The second possibility is that you actually are just hitting the memory limit on your machine - but don't worry, the corpus file isn't actually 16 GB in size!

Regardless of which issue is leading to the error you're seeing, the root cause is in the SpaCy dependency. SpaCy 2.0, which we upgraded to last year, contains a lot of really useful changes but also introduces an issue with memory usage on large datasets (see this github issue). This causes the spacy_objs object to eat much more memory than it really needs to, and by extension causes the pickled object to be much larger than needed. Thankfully, there is a workaround for this, which we have implemented and will be rolling out in the 2.0 release.

**Comment by calebchiam** [comment]
Hi! This has been fixed as of the release of ConvoKit 2.0 a little over a month ago. We no longer include tennis_question_typology.py in the repo as an example (because of redundancy), but the [streamlined Transformer interface](https://zissou.infosci.cornell.edu/socialkit/documentation/architecture.html) of QuestionTypology makes it easy to apply the QuestionTypology extraction to any corpus.

If you are still interesting in running tennis_question_typology.py, this updated code will do the job:

```
import os
import pkg_resources
import numpy as np

from convokit import Corpus, QuestionTypology, download

#Load the corpus
num_clusters = 8
corpus = Corpus(filename=download("tennis-corpus"))

#Extract clusters of the motifs and assign questions to these clusters
qt = QuestionTypology(num_dims=25,
  num_clusters=num_clusters, verbose=False, random_seed=125)

qt.fit(corpus)
qt.transform(corpus)

qt.display_totals()
print('10 examples for types 1-8:')
for i in range(num_clusters):
    qt.display_motifs_for_type(i, num_egs=10)
    qt.display_answer_fragments_for_type(i, num_egs=10)
    qt.display_question_answer_pairs_for_type(corpus, i, num_egs=10)
```

Hope this was helpful, and thanks for raising the issue.
